src/main/kotlin/me/zeroeightsix/kami/util/MidiParser.kt

```diff
@@ -0,0 +1,88 @@
+package me.zeroeightsix.kami.util
+
+import java.io.File
+import java.io.IOException
+import java.util.ArrayList
+import java.util.TreeMap
+import javax.sound.midi.InvalidMidiDataException
+import javax.sound.midi.MetaMessage
+import javax.sound.midi.MidiEvent
+import javax.sound.midi.MidiSystem
+import javax.sound.midi.Sequence
+import javax.sound.midi.ShortMessage
+
+object MidiParser {
+    const val SET_TEMPO = 0x51
+    const val TIME_SIGNATURE = 0x58
+    const val PITCH_CHANGE = 0xC0
+    const val QUARTER_NOTE_IN_MICROSECONDS = 0x07A120
+
+    /**
+     * Takes a midi file as a string, parses it, and returns what notes play at times in a Map.
+     *
+     * @param filename - Midi File
+     * @return - Notes that play at what time
+     * @throws InvalidMidiDataException - If the File is corrupt
+     * @throws IOException              - If the file isn't found
+     */
+    @Throws(InvalidMidiDataException::class, IOException::class)
+    fun parse(filename: String?, maxChannels: Int): TreeMap<Long, ArrayList<Note>> {
```

Comment:
filename should be nonnull:
```suggestion
    fun parse(filename: String, maxChannels: Int): TreeMap<Long, ArrayList<Note>> {
```